### ISSUE TITLE
[Nightly] Low map zoom is always reset to specific zoom level

### ISSUE THREAD

**Issue by MagpieFourtyTwo** [cgeo/cgeo#9332]
**Describe the bug:**
A short while ago we had the problem that zoom levels >100 m (i. e. 200 m, 500 m, 1 km ...) were reset to 100 m in subsequent map calls, which has been corrected with #9293. But while this is working perfectly fine, it's now (or perhaps even then?) the other way round: zoom levels <100 m (i. e. 50 m, 20 m, 10 m ...) are reset to 100 m in subsequent OSM map calls ...

In GM it's nearly the same, only difference: in GM and GSat the magic limit is even 200 m.

**To Reproduce:**
1. open any map view (no matter if Live map or cache map)
2. zoom until the scale shows "50 m"
3. leave map
4. re-enter map

**Actual behavior/state after performing these steps:**
If OSM was selected, zoom scale will be reset to 100 m, with GM/GSat it will even be reset to 200 m

**Expected behavior/state after performing these steps:**
Zoom should be constant over subsequent calls. 

**Version of c:geo used:**
2020.11.01-NB-cd04617

**Is the problem reproducible:**
Yes

**Comment by eddiemuc** [comment]
Indeed there is an initial maximal zoom limit set to 16 (which is about 100m on the scale). I set it up to 25 (which in practice means that there is no limit at all any more). PR is #9333

**Comment by Lineflyer** [comment]
Was this introduced only recently by your change, because AFAIR we implemented an upper limit on purpose (for live map) to avoid meaningless requests for a whole country. 
If it was in before it was probably implicitly changed by the correction to align OSM and Google. Could it then be set to +3 which means 19?

**Comment by Lineflyer** [comment]
Forget my previous comment. I was thinking the other way around.

**Comment by eddiemuc** [comment]
> Was this introduced only recently by your change, because AFAIR we implemented an upper limit on purpose (for live map) to avoid meaningless requests for a whole country.
> If it was in before it was probably implicitly changed by the correction to align OSM and Google. Could it then be set to +3 which means 19?

@Lineflyer 
Before my previous change, the limit 16 was used de facto as an upper-bound (meaning: no zoom HIGHER than 16 was saved). I turned this around, but I did not realized at the time that now (as @MagpieFourtyTwo discovered) no zoom LOWER than 16 could be saved :-)
If you introduced this boundary in the past on purpose to avoid meaningless requests for a whole country: this boundary is currently effectively removed. Even if you zoom to the whole world, then leave the map and return, the map will start with the whole world again.
Should I restore the upper bound? Where should I then set it? 16 seems a bit harsh, maybe 10?

**Comment by Lineflyer** [comment]
Understood.
I dont know...as I dont know what area is covered with zoom=10. Lets try, we can adjust later.

**Comment by MagpieFourtyTwo** [comment]
@eddiemuc What do you mean with "I turned this around"?  The bounds or the complete level "naming"?

It was @moving-bits IIRC who once posted some examples for the relation of zoom level and shown map scale.
IIRC, 0 was world level and 25 s/th near to reality ;)  Following this I would guess the values below.
Or is it this level numbering, you turned around completely ...?

OSM Level|metric|imperial
--|--|--
0|10.000 km|5.000 mi
1|5.000 km|5.000 mi
2|2.000 km|2.000 mi
3|2.000 km|1.000 mi
4|1.000 km|500 mi
5|500 km|200 mi
6|200 km|100 mi
7|100 km|50 mi
8|50 km|20 mi
9|20 km|20 mi
10|10 km|10 mi
11|5 km|5 mi
12|2 km|2 mi
13|2 km|1 mi
14|1 km|2000 ft
15|500 m|1000 ft
16|200 m|500 ft
17|100 m|200 ft
18|50 m|200 ft
19|20 m|100 ft
20|10 m|50 ft
21|5 m|20 ft
22|2 m|10 ft

GM Level|metric
--|--
0|2000 km
1|1000 km
2|600 km
3|300 km
4|100 km
5|80 km
6|40 km
7|20 km
8|10 km
9|5 km
10|2 km
11|1 km
12|600 m
13|300 m
14|100 m
15|80 m
16|40 m
17|20 m
18|10 m

@Lineflyer What I currently don't get is, why there should be an upper limit at all, as there will be just 500 caches, anyway ...

**Comment by MagpieFourtyTwo** [comment]
Ähem, for my understanding: Got the values by zooming out to max (world level) and then zooming in step by step ... and right now I recognize, the difference between OSM and GM level is about 3, most of the times. E. g. on OSM 1 km is given at level 14 and in GM at level 11 ... are these the magic 3 ...? And why are they currently not needed anymore?

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo : I try to clear up my fuzzy statement "I turned this around" in the following.

When I investigated #9288 ("preserve zoom level") a few days ago I found in code that whenever the map was opened and the initial zoom level was retrieved from Settings, this value was set to 16 if it was lower than 16 (on this level, OSM displays a scale of "100m"). This resulted in the behaviour described in the issue: previous zoom was not restored if map was zoomed out more than this.

I was puzzled why anyone would hard-code this behaviour, so I asked for a close PR review (see https://github.com/cgeo/cgeo/issues/9288#issuecomment-717396428 and the coresponding PR #9298). At that time I didn't know, now I know from @Lineflyer s previous comment on this thread: obviously this was done to prevent loading too many caches on map opening (when displayeed map area is large, more caches are expected on it)

Still for #9288 I changed the behaviour, but made a mistake myself. I removed the logic restricting zoom level on load, but introduced the opposite logic: now zoom levels are hard-coded to 16 if they are HIGHER than 16 (resulting in the behaviour you described in this issue).

I created PR #9333 which sets the boundary  for high zoom levels to 25 (which is never reached). I will now also include a new low boundary and set it to 10 to deal with the "caching loading" issue (as discussed above with @Lineflyer)

**Comment by eddiemuc** [comment]
This is a view with map zoom 10 in OSM:
![image](https://user-images.githubusercontent.com/6909759/97848200-05817b80-1cf1-11eb-98fc-05e18eb08b3d.png)

**Comment by Lineflyer** [comment]
> This is a view with map zoom 10 in OSM:

IMHO this should be OK as starting zoom if the user previously zoomed out below that.

**Comment by MagpieFourtyTwo** [comment]
> IMHO this should be OK as starting zoom if the user previously zoomed out below that.

Sorry, but still don't get why there should be a max zoom at all ...?  For me such a behavior would not really be comprehensible. And really annoying. If I (or esp. Joe Average) look at a map in full (resp. more than 10 km) zoom, they for sure have reasons for it and would not really be happy if the view will every time be reset to a way closer level and have to zoom out again, just because they shortly left the view ...

I of course understand the factor of server load, but a) the amount of caches is limited to 500 anyway, and b) why not just limit the rectangle for the server query?  And if a rectangle (which may even correspond to zoom level 10, thus 10 km / 5 mi) results in less than let's say 400 caches, it may even be expanded until a reasonable amount of caches is reached ...

And, BTW, in such cases a toast would be really helpful, i. e. s/th like "There may be more caches in this area, but request is limited to 500 caches - please zoom in to see more."

**Comment by fm-sys** [comment]
![Screenshot_20201102_101805_cgeo geocaching_copy_270x585](https://user-images.githubusercontent.com/64581222/97851801-24ced780-1cf6-11eb-8c1f-25080472a280.jpg)

IMHO this would be a good starting point for maximum zooming in as zooming in even more doesn't really make sense... And I would also vote for doesn't have a maximum out zoom limit

**Comment by Lineflyer** [comment]
> And I would also vote for doesn't have a maximum out zoom limit

OK, then lets drop the zoom out limit. Maybe this would also bring other interworking problems, e.g. when mapping a cache whose waypoints are far spread.

@MagpieFourtyTwo 
I suggest not to mix topics in here. Feel free to open dedicated issues for other improvement proposals. Don't get me wrong, we can discuss anything, but lets really stay on topic in one issue otherwise its hard to follow the separate ideas.

**Comment by MagpieFourtyTwo** [comment]
@Lineflyer 
> I suggest not to mix topics in here.

Never mind, the thought just crossed my mind and just wanted to know if someone is immediately yelling "NOOOOO!"  ;)
Separate issue resp. FR will of course follow, as usual.  ;)
